File: bms/go/go-gt/Hungarian/Hungarian.go
package test

import (
	"math"
)

var ()

const ()

func Hungarian(g *Matrix) (xy []int64, yx []int64) {
	e := initH(g)
	e.augment()
	return e.Xy, e.Yx
}

func initH(g *Matrix) (e *Env) {
	var i, j int64
	e = newEnv(g.N)
	e.G = g
	e.N = g.N
	for i = 0; i < e.N; i++ {
		for j = 0; j < e.N; j++ {
			e.Lx[i] = max(e.Lx[i], e.G.Get(i, j))
		}
	}
	return e
}

func newEnv(n int64) *Env {
	e := new(Env)
	e.M = 0
	e.N = n
	e.T = make([]bool, n)
	e.S = make([]bool, n)
	e.Slack = make([]int64, n)
	e.Slackx = make([]int64, n)
	e.Prev = make([]int64, n)
	e.Xy = make([]int64, n)
	e.Yx = make([]int64, n)
	e.Lx = make([]int64, n)
	e.Ly = make([]int64, n)
	var i int64
	for i = 0; i < n; i++ {
		e.Xy[i] = -1
		e.Yx[i] = -1
	}
	return e
}

func (m Matrix) Get(i int64, j int64) int64 {
	return m.A[i*m.N+j]
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func (e *Env) augment() {
	var i, j, wr, rd, r int64
	wr = 0
	rd = 0
	q := make([]int64, e.N)
	if e.M == e.N {
		return
	}
	for i = 0; i < e.N; i++ {
		if e.Xy[i] == -1 {
			wr++
			q[wr] = i
			r = i
			e.Prev[i] = -2
			e.S[i] = true
			break
		}
	}
	for i = 0; i < e.N; i++ {
		e.Slack[i] = e.Lx[r] + e.Ly[i] - e.G.Get(r, i)
		e.Slackx[i] = r
	}
	for {
		for rd < wr {
			rd++
			i = q[rd]
			for j = 0; j < e.N; j++ {
				if e.G.Get(i, j) == e.Lx[i]+e.Ly[j] && !e.T[j] {
					if e.Yx[j] == -1 {
						break
					}
					e.T[j] = true
					wr++
					q[wr] = e.Yx[j]
					e.add(e.Yx[j], i)
				}
			}
			if j < e.N {
				break
			}
		}
		if j < e.N {
			break
		}
		e.update()
		wr = 0
		rd = 0
		for j = 0; j < e.N; j++ {
			if !e.T[j] && e.Slack[j] == 0 {
				if e.Yx[i] == -1 {
					i = e.Slackx[j]
					break
				} else {
					e.T[j] = true
					if !e.S[e.Yx[j]] {
						wr++
						q[wr] = e.Yx[j]
						e.add(e.Yx[j], e.Slackx[j])
					}
				}
			}
		}
		if j < e.N {
			return
		}
	}
	if j < e.N {
		e.M++
		for i != -2 {
			k := e.Xy[i]
			e.Yx[j] = i
			e.Xy[i] = j
			i = e.Prev[i]
			j = k
		}
		e.augment()
	}
}

func (e *Env) add(i, p int64) {
	var j int64
	e.S[i] = true
	e.Prev[i] = p
	for j = 0; j < e.N; j++ {
		if e.Lx[i]+e.Ly[i]-e.G.Get(i, j) < e.Slack[i] {
			e.Slack[i] = e.Lx[i] + e.Ly[i] - e.G.Get(i, j)
			e.Slackx[i] = j
		}
	}
}

func (e *Env) update() {
	var i int64
	var d int64 = math.MaxInt64
	for i = 0; i < e.N; i++ {
		if !e.T[i] {
			d = min(d, e.Slack[i])
		}
	}
	for i = 0; i < e.N; i++ {
		if e.S[i] {
			e.Lx[i] -= d
		}
	}
	for i = 0; i < e.N; i++ {
		if e.T[i] {
			e.Ly[i] += d
		}
	}
	for i = 0; i < e.N; i++ {
		if !e.T[i] {
			e.Slack[i] -= d
		}
	}
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

type Matrix struct {
	N int64
	A []int64
}

type Env struct {
	M, N   int64
	G      *Matrix
	T, S   []bool
	Slack  []int64
	Slackx []int64
	Prev   []int64
	Xy, Yx []int64
	Lx, Ly []int64
}

File: translations_only/fail/go/go-gt/Hungarian/claude2/BaseRepair/Hungarian.rs


use std::cmp::{max, min}; 

#[derive(Clone)] 
struct Matrix {
    n: i64,
    a: Vec<i64>,
}

fn get(m: &Matrix, i: i64, j: i64) -> i64 {
    m.a[i as usize * m.n as usize + j as usize]  
}

struct Env {
    m: i64,
    n: i64, 
    g: Box<Matrix>,
    t: Vec<bool>,
    s: Vec<bool>,
    slack: Vec<i64>,
    slackx: Vec<i64>,
    prev: Vec<i64>,
    xy: Vec<i64>,
    yx: Vec<i64>,
    lx: Vec<i64>,
    ly: Vec<i64>,
}

fn new_env(n: i64) -> Env {
    Env {
        m: 0,
        n,
        g: Box::new(Matrix { n: 0, a: vec![] }),
        t: vec![false; n as usize],
        s: vec![false; n as usize],
        slack: vec![-1; n as usize],
        slackx: vec![-1; n as usize],
        prev: vec![-1; n as usize],
        xy: vec![-1; n as usize],
        yx: vec![-1; n as usize],
        lx: vec![-1; n as usize],
        ly: vec![-1; n as usize],
    }
}

fn init_h(g: &Matrix) -> Env {
    let mut e = new_env(g.n);
    e.g = Box::new(g.clone());
    e.n = g.n;

    for i in 0..e.n {
        for j in 0..e.n {
            e.lx[i as usize] = max(e.lx[i as usize], get(g, i, j));
        }
    }

    e
}

fn hungarian(g: &Matrix) -> (Vec<i64>, Vec<i64>) {
    let mut e = init_h(&g);
    augment(&mut e);
    (e.xy, e.yx)  
}

fn augment(e: &mut Env) {
    let mut i: i64;
    let mut j: i64 = 0;
    let mut wr: i64 = 0;
    let mut rd: i64 = 0;
    let mut r: i64 = 0;
    let mut q = vec![-1; e.n as usize];

    if e.m == e.n {
        return;
    }

    'outer: for i in 0..e.n {
        if e.xy[i as usize] == -1 {
            wr += 1;
            q[wr as usize] = i;
            r = i;
            e.prev[i as usize] = -2;
            e.s[i as usize] = true;
            break;
        }
    }

    for i in 0..e.n {
        e.slack[i as usize] = e.lx[r as usize] + e.ly[i as usize] - get(e.g.as_ref(), r, i);
        e.slackx[i as usize] = r;
    }

    'inner: loop {
        for rd in 0..e.n {
            i = q[rd as usize];

            for j in 0..e.n {
                if get(e.g.as_ref(), i, j) == e.lx[i as usize] + e.ly[j as usize] && !e.t[j as usize] {
                    if e.yx[j as usize] == -1 {
                        break 'inner;
                    }

                    e.t[j as usize] = true;
                    wr += 1;
                    q[wr as usize] = e.yx[j as usize];
                    add(e, e.yx[j as usize], i);
                }
            }

            if j < e.n {
                break; 
            }
        }

        if j < e.n {
            break;
        }

        update(e);

        wr = 0;
        rd = 0;

        for j in 0..e.n {
            if !e.t[j as usize] && e.slack[j as usize] == 0 {
                if e.yx[j as usize] == -1 {
                    i = e.slackx[j as usize];
                    break;
                } else {
                    e.t[j as usize] = true;

                    if !e.s[e.yx[j as usize] as usize] {
                        wr += 1;
                        q[wr as usize] = e.yx[j as usize];
                        add(e, e.yx[j as usize], e.slackx[j as usize]);
                    }
                }
            }
        }

        if j < e.n {
            return;
        }
    }

    if j < e.n {
        e.m += 1;

        let mut i = -2;
        while i != -2 {
            let k = e.xy[i as usize];
            e.yx[j as usize] = i;
            e.xy[i as usize] = j;
            i = e.prev[i as usize];
            j = k;
        }

        augment(e);
    }
}

fn add(e: &mut Env, i: i64, p: i64) {
    e.s[i as usize] = true;
    e.prev[i as usize] = p;

    for j in 0..e.n {
        if e.lx[i as usize] + e.ly[i as usize] - get(e.g.as_ref(), i, j) < e.slack[i as usize] {
            e.slack[i as usize] = e.lx[i as usize] + e.ly[i as usize] - get(e.g.as_ref(), i, j);
            e.slackx[i as usize] = j;
        }
    }
}

fn update(e: &mut Env) {
    let mut i: i64;
    let mut d = i64::MAX;

    for i in 0..e.n {
        if !e.t[i as usize] {
            d = min(d, e.slack[i as usize]);
        }
    }

    for i in 0..e.n {
        if e.s[i as usize] {
            e.lx[i as usize] -= d;
        }
    }

    for i in 0..e.n {
        if e.t[i as usize] {
            e.ly[i as usize] += d;
        }
    }

    for i in 0..e.n {
        if !e.t[i as usize] {
            e.slack[i as usize] -= d;
        }
    }
}

